atomdb/base.py

```diff
@@ -318,6 +424,194 @@ def __get__(self, obj: T, cls: Optional[Type[T]] = None):
         raise NotImplementedError
 
 
+def generate_getstate(cls: Type["Model"], include_defaults: bool = True) -> GetStateFn:
+    """Generate an optimized __getstate__ function for the given model.
+
+    Parameters
+    ----------
+    cls: Type[Model]
+        The clase to generate a getstate function for.
+
+    Returns
+    -------
+    result: GetStateFn
+        A function optimized to generate the state for the given model class.
+
+    """
+    template = [
+        "def __getstate__(self, scope=None):",
+        "scope = scope or {}",
+        "scope[self.__ref__] = self",
+        "state = {",
+    ]
+    default_flatten = cls.serializer.flatten
+    members = cls.members()
+    namespace = {
+        "default_flatten": default_flatten,
+    }
+    for f in cls.__fields__:
+        # Since f is potentially an untrusted input, make sure it is a valid
+        # python identifier to prevent unintended code being generated.
+        if not f.isidentifier():
+            raise ValueError(f"Field '{f}' cannot be used for code generation")
+        m = members[f]
+        meta = m.metadata or {}
+        flatten = meta.get("flatten", default_flatten)
+        if flatten is default_flatten:
+            if is_primitive_member(m):
+                expr = f"self.{f}"
+            else:
+                expr = f"default_flatten(self.{f}, scope)"
+        else:
+            namespace[f"flatten_{f}"] = flatten
+            expr = f"flatten_{f}(self.{f}, scope)"
+        template.append(f'    "{f}": {expr},')
+
+    if include_defaults:
+        template.append('    "__model__": self.__model__,')
+        template.append('    "__ref__": self.__ref__,')
+        template.append("}")
+        template.append("if self._id is not None:")
+        template.append('    state["_id"] = self._id')
+    else:
+        template.append("}")
+    template.append("return state")
+    source = "\n    ".join(template)
+    return generate_function(source, namespace, "__getstate__")
+
+
+def generate_restorestate(cls: Type["Model"]) -> RestoreStateFn:
+    """Generate an optimized __restorestate__ function for the given model.
+
+    Parameters
+    ----------
+    cls: Type[Model]
+        The clase to generate a getstate function for.
+
+    Returns
+    -------
+    result: RestoreStateFn
+        A function optimized to restore the state for the given model class.
+
+    """
+    # Python must do some caching because using key in state and state[key]
+    # seems to be faster than using get
```

Comment:
It may be because get rely on handling the KeyError

Reply:
Thanks for the comment, I didn't know that, but that makes sense. Will be interesting if this changes with 3.11's "Zero cost" exceptions (https://docs.python.org/3.11/whatsnew/3.11.html#optimizations)

Reply:
The unhappy path will still be costlier than you check, it will depend on the likeliness of the key missing.